Sources/SentryCrash/Recording/Tools/SentryHook.c

```diff
@@ -56,6 +28,80 @@ sentrycrash__async_backtrace_decref(sentrycrash_async_backtrace_t *bt)
     }
 }
 
+/**
+ * This is a poor-mans concurrent hashtable.
+ * We have N slots, using modulo of the thread ID. Using atomic load / compare-exchange to make sure
+ * that the slot indeed belongs to the thread we want to work with.
+ */
+
+#define SENTRY_MAX_ASYNC_THREADS (128 - 1)
+
+typedef struct {
+    SentryCrashThread thread;
+    sentrycrash_async_backtrace_t *backtrace;
+} sentrycrash_async_caller_t;
+
+static sentrycrash_async_caller_t sentry_async_callers[SENTRY_MAX_ASYNC_THREADS] = { 0 };
+
+static size_t
+sentrycrash__thread_idx(SentryCrashThread thread)
+{
+    return (thread * 0x100000001b3 + 0xcbf29ce484222325) % SENTRY_MAX_ASYNC_THREADS;
+}
+
+sentrycrash_async_backtrace_t *
+sentrycrash_get_async_caller_for_thread(SentryCrashThread thread)
+{
+    if (!__atomic_load_n(&hooks_active, __ATOMIC_RELAXED)) {
+        return NULL;
+    }
+
+    size_t idx = sentrycrash__thread_idx(thread);
+    sentrycrash_async_caller_t *caller = &sentry_async_callers[idx];
+    if (__atomic_load_n(&caller->thread, __ATOMIC_ACQUIRE) == thread) {
+        sentrycrash_async_backtrace_t *backtrace = caller->backtrace;
+        // we read the thread id *again*, if it is still the same, the backtrace pointer we
+        // read in between is valid
+        if (__atomic_load_n(&caller->thread, __ATOMIC_SEQ_CST) == thread) {
+            return backtrace;
+        }
+    }
+    return NULL;
+}
+
+static void
+sentrycrash__set_async_caller(sentrycrash_async_backtrace_t *backtrace)
+{
+    SentryCrashThread thread = sentrycrashthread_self();
+
+    size_t idx = sentrycrash__thread_idx(thread);
+    sentrycrash_async_caller_t *caller = &sentry_async_callers[idx];
+
+    SentryCrashThread expected = (SentryCrashThread)NULL;
+    bool success = __atomic_compare_exchange_n(
+        &caller->thread, &expected, thread, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+
+    if (success) {
+        __atomic_store_n(&caller->backtrace, backtrace, __ATOMIC_RELEASE);
+    }
```

Comment:
Does this need to be done atomically? Trying to follow the lock-free logic here but it seems like this `success` check could cause a race depending on how things get scheduled?